Fix a bug in FileDatabase::DbGetClassProperty (#816)

bourtemb opened the pull request:
Add documentation to FileDatabase::DbGetClassProperty method
Translate some comments from Italian to English

bourtemb committed:
Fix a bug in FileDatabase::DbGetClassProperty (#816)
Add documentation to FileDatabase::DbGetClassProperty method
Translate some comments from Italian to English

mliszcz commented on cppapi/client/filedatabase.cpp:
```diff
@@ -1612,8 +1635,8 @@ CORBA::Any*   FileDatabase :: DbGetClassProperty(CORBA::Any& send)
-			// m_server.classes[i] e' la classe che cerchiamo
-			for (unsigned int j = 1; j < (*data_in).length(); j++)  // in 0 c'e' il nome della classe e poi a seguire le proprieta
+			// m_server.classes[i] is the class we are looking for
+			for (unsigned int j = 1; j < (*data_in).length(); j++)  // at index 0 is the name of the class, property names are following
 			{
 				unsigned long nb_prop_defined = m_server.classes[i]->properties.size();
```
Consider adding a variable that would hold property name for the current loop iteration.
```cpp
auto const &prop_name = (*data_in)[j];
```
This `(*data_in)[j]` is repeated three times below. Not only this is unreadable and hard to maintain but also was the source of this bug.

mliszcz commented on cppapi/client/filedatabase.cpp:
```diff
@@ -1642,11 +1665,13 @@ CORBA::Any*   FileDatabase :: DbGetClassProperty(CORBA::Any& send)
-					(*data_out)[index] = Tango::string_dup((*data_in)[i + 1].in());index++;
+					// The requested property name is returned, followed by a 0,
+					// meaning the length of this property value is 0 ( <=> class property not found )
+					(*data_out)[index] = Tango::string_dup((*data_in)[j].in());index++;
```
This `index++` scattered all over the place is not only ugly but dangerous if one forgets to add it. There are few options that we could use to avoid that, at least these come to my mind:

1. Have a lambda that increases this index for us. But this could only be done in 9.4.
    ```cpp
    auto const push_data_out = [index = 0u](auto const &elem) mutable
    {
        (*data_out)[index] = Tango::string_dup(elem);
        index++;
    };
    ```
    Here we could simply call:
    ```cpp
    push_data_out(prop_name);
    push_data_out("0");
    ```

2. Collect all data in a vector and then copy that into the sequence. Note that we want to copy only the pointers. And it's not exception-safe but this is non-throwing code.
    ```cpp
    std::vector<char*> output{};
    output.push_back(Tango::string_dup("whatever"));

    auto *result = new DevVarStringArray{};
    result.length(output.size());
    std::memcpy(result.get_buffer(), output.data(), output.size() * sizeof(char*));
    ```

3. Write some utility classes which will provide a sane API for CORBA sequences. Maybe push_back like vector has, iteration support like in std;:span, etc. But this is something that we may rather consider for 9.4, not for backports.

mliszcz reviewed (commented):
I have some general comments regarding the implementation of this function. Maybe we could use this opportunity to do a small cleanup here. Github does not allow to put comments on lines not touched by the PR, so I'll leave them here:

https://github.com/tango-controls/cppTango/blob/1ddda45603e02ac279e9da6dd88e05da95542b61/cppapi/client/filedatabase.cpp#L1624
Unneeded and also incorrect because earlier we assign `num_prop = data_in->length() - 1;`. This variable should be declared const.

https://github.com/tango-controls/cppTango/blob/1ddda45603e02ac279e9da6dd88e05da95542b61/cppapi/client/filedatabase.cpp#L1629-L1633
Multiple problems here but main one is that `value` is a vector, c++ requires `size()` to return unsigned integer, printing with `%d` is at least suspicious. I haven't checked printf implementation but I guess there is a reason why they have `%u`. I'd replace this with a simple:
```
sprintf(num_vals_str, "%lu", static_cast<unsigned long>(num_val))
```

t-b commented:
@mliszcz This PR targets 9.3-backports so I would favour the smallest possible bugfix. Your cleanup is IMHO good for tango-9-lts. Regrading the snprintf thing I have something in https://github.com/tango-controls/cppTango/pull/792/files#diff-bdc1f56f72d7a12798a63b803aebf5a6b49233a15c0cdefb705572c03d23f566 which deals with that. The PR is not yet done though.

mliszcz commented:
@t-b sure, these are just comments, Reynald may not agree with them or drop or postpone them, that's how reviews work. I put them here mainly to allow us re-consider these changes when forwardport PR comes. Otherwise I'd probably forget what I learnt about this code today.

bourtemb committed:
FileDatabase::DbGetClassProperty: num_prop variable is now const
num_prop local variable in FileDatabase::DbGetClassProperty method is
now declared const.
Removed an uneeded and incorrect incrementation of this variable too.
Thanks @mliszcz for the suggestion.

bourtemb commented:
> I have some general comments regarding the implementation of this function. Maybe we could use this opportunity to do a small cleanup here. Github does not allow to put comments on lines not touched by the PR, so I'll leave them here:
> 
> https://github.com/tango-controls/cppTango/blob/1ddda45603e02ac279e9da6dd88e05da95542b61/cppapi/client/filedatabase.cpp#L1624
> 
> Unneeded and also incorrect because earlier we assign `num_prop = data_in->length() - 1;`. This variable should be declared const.

Indeed. Thanks @mliszcz . These suggestions should be implemented in 1bd6d38f902a8dd6195870dd3042d0fdfea2911f
